client/src/main/scala/org/http4s/client/middleware/Retry.scala

```diff
@@ -2,71 +2,141 @@ package org.http4s
 
 object Retry {
 
   private[this] val logger = getLogger
 
-  private[this] val RetriableStatuses = Set(
-    RequestTimeout,
-    // TODO Leaving PayloadTooLarge out until we model Retry-After
-    InternalServerError,
-    ServiceUnavailable,
-    BadGateway,
-    GatewayTimeout
-  )
-
-  def apply(backoff: Int => Option[FiniteDuration])(client: Client)(implicit ec: ExecutionContext, scheduler: Scheduler): Client = {
+  def apply(policy: RetryPolicy)(client: Client)(implicit ec: ExecutionContext, scheduler: Scheduler): Client = {
     implicit val s: Strategy = Strategy.fromExecutionContext(ec)
 
     def prepareLoop(req: Request, attempts: Int): Task[DisposableResponse] = {
       client.open(req).attempt flatMap {
         // TODO fs2 port - Reimplement request isIdempotent in some form
-        case Right(dr @ DisposableResponse(Response(status, _, _, _, _), _)) if RetriableStatuses(status) =>
-          backoff(attempts) match {
+        case right @ Right(dr @ DisposableResponse(response, _)) =>
+          policy(req, Right(dr.response), attempts) match {
             case Some(duration) =>
-              logger.info(s"Request ${req} has failed on attempt #${attempts} with reason ${status}. Retrying after ${duration}.")
-              dr.dispose.flatMap(_ => nextAttempt(req, attempts, duration))
+              logger.info(s"Request ${req} has failed on attempt #${attempts} with reason ${response.status}. Retrying after ${duration}.")
+              dr.dispose.flatMap(_ => nextAttempt(req, attempts, duration, response.headers.get(`Retry-After`)))
             case None =>
-              logger.info(s"Request ${req} has failed on attempt #${attempts} with reason ${status}. Giving up.")
               Task.now(dr)
           }
-        case Right(dr) =>
-          Task.now(dr)
         case Left(e) =>
-          backoff(attempts) match {
+          policy(req, Left(e), attempts) match {
             case Some(duration) =>
               logger.error(e)(s"Request ${req} threw an exception on attempt #${attempts} attempts. Retrying after ${duration}.")
-              nextAttempt(req, attempts, duration)
+              nextAttempt(req, attempts, duration, None)
             case None =>
-              // info instead of error(e), because e is not discarded
-              logger.info(s"Request ${req} threw an exception on attempt #${attempts} attempts. Giving up.")
               Task.fail(e)
           }
       }
     }
 
-    def nextAttempt(req: Request, attempts: Int, duration: FiniteDuration): Task[DisposableResponse] = {
-      // TODO honor Retry-After header
-      prepareLoop(req.withEmptyBody, attempts + 1).schedule(duration)
+    def nextAttempt(req: Request, attempts: Int, duration: FiniteDuration, retryHeader: Option[`Retry-After`]): Task[DisposableResponse] = {
+      val headerDuration = retryHeader.map { h =>
+        h.retry match {
+          case Left(d) => Instant.now().until(d.toInstant, ChronoUnit.SECONDS)
```

Comment:
This is the decreasing time location?

Reply:
Yeah, if a Retry-After is a Left, we need to calculate the delta from now until then.  The Right is expressed as a wait time in seconds.